File: मम/oopcg-se-25/gB_BankingSystem.cpp
#include <iostream>
#include <iomanip>
#include <limits>
#include <string>
#include <vector>
#include <ctime>

struct Transaction {
    std::time_t ts;
    std::string type;
    double amount;
    double balance_after;
};

class Account {
  private:
    static constexpr double kDailyLimit = 5000.0; // Daily withdrawal limit
    int account_no_;
    std::string holder_name_;
    double balance_{};
    std::vector<Transaction> passbook_;

    double withdrawn_today_{};
    std::time_t last_withdraw_day_{};

    static int generateAccountNo() {
        static int next = 1001;
        return next++;
    }

    void addTransaction(const std::string &type, double amount) {
        passbook_.push_back({std::time(nullptr), type, amount, balance_});
    }

    static bool isSameDay(std::time_t a, std::time_t b) {
        std::tm tm_a = *std::localtime(&a);
        std::tm tm_b = *std::localtime(&b);
        return tm_a.tm_year == tm_b.tm_year && tm_a.tm_yday == tm_b.tm_yday;
    }

  public:
    Account(std::string name, double initial) : account_no_(generateAccountNo()), holder_name_(std::move(name)), balance_(initial) {
        addTransaction("OPEN", initial);
    }

    void deposit(double amt) {
        balance_ += amt;
        addTransaction("DEPOSIT", amt);
    }

    bool withdraw(double amt) {
        std::time_t now = std::time(nullptr);
        if (!isSameDay(now, last_withdraw_day_)) {
            withdrawn_today_ = 0;
            last_withdraw_day_ = now;
        }

        if (amt > balance_) {
            std::cerr << "Insufficient balance.\n";
            return false;
        }
        if (withdrawn_today_ + amt > kDailyLimit) {
            std::cerr << "Daily withdrawal limit exceeded.\n";
            return false;
        }

        balance_ -= amt;
        withdrawn_today_ += amt;
        addTransaction("WITHDRAW", amt);
        return true;
    }

    double balance() const { return balance_; }

    void printInfo() const {
        std::cout << "Account No : " << account_no_ << "\nName       : " << holder_name_ << "\nBalance    : $" << std::fixed << std::setprecision(2) << balance_ << "\n";
    }

    void printPassbook() const {
        std::cout << "\n--- Passbook ---\n";
        for (const auto &t : passbook_) {
            std::tm tm = *std::localtime(&t.ts);
            char buf[25];
            std::strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", &tm);
            std::cout << buf << " | " << std::setw(8) << t.type << " | " << std::setw(10) << t.amount << " | Bal: " << t.balance_after << "\n";
        }
    }

    void printPassbook(std::time_t from, std::time_t to) const {
        std::cout << "\n--- Passbook from to ---\n";
        for (const auto &t : passbook_) {
            if (t.ts >= from && t.ts <= to) {
                std::tm tm = *std::localtime(&t.ts);
                char buf[25];
                std::strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", &tm);
                std::cout << buf << " | " << std::setw(8) << t.type << " | " << std::setw(10) << t.amount << " | Bal: " << t.balance_after << "\n";
            }
        }
    }
};

static void clearCin() {
    std::cin.clear();
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
}

int main() {
    Account *account = nullptr;
    int option;
    while (true) {
        std::cout << "\n===== Banking System =====\n";
        std::cout << "1. Create account\n2. Deposit\n3. Withdraw\n4. Check balance\n5. Display account info\n6. Passbook print\n0. Exit\nSelect: ";
        if (!(std::cin >> option)) {
            clearCin();
            continue;
        }

        if (option == 0)
            break;

        if (!account && option != 1) {
            std::cout << "Please create an account first.\n";
            continue;
        }

        switch (option) {
            case 1: {
                std::string name;
                double initial;
                std::cout << "Enter name: ";
                clearCin();
                std::getline(std::cin, name);
                std::cout << "Initial deposit: $";
                std::cin >> initial;
                account = new Account(name, initial);
                std::cout << "Account created successfully!\n";
                break;
            }
            case 2: {
                double amt;
                std::cout << "Deposit amount: $";
                std::cin >> amt;
                account->deposit(amt);
                std::cout << "Deposited successfully.\n";
                break;
            }
            case 3: {
                double amt;
                std::cout << "Withdraw amount: $";
                std::cin >> amt;
                account->withdraw(amt);
                break;
            }
            case 4: {
                std::cout << "Current balance: $" << std::fixed << std::setprecision(2) << account->balance() << "\n";
                break;
            }
            case 5: {
                account->printInfo();
                break;
            }
            case 6: {
                std::cout << "Print full passbook? (y/n): ";
                char yn;
                std::cin >> yn;
                if (yn == 'y' || yn == 'Y') {
                    account->printPassbook();
                } else {
                    std::cout << "Enter 'from' timestamp (epoch): ";
                    std::time_t from, to;
                    std::cin >> from;
                    std::cout << "Enter 'to' timestamp (epoch): ";
                    std::cin >> to;
                    account->printPassbook(from, to);
                }
                break;
            }
            default:
                std::cout << "Invalid option.\n";
        }
    }

    delete account;
    return 0;
}

File: मम/oopcg-se-25/gB_BankingSystem.java
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Banking system demonstration written in Java, mirroring the Group-C C++
 * assignment but fully object-oriented in Java.
 */
public class gB_BankingSystem {

    private static final Scanner scanner = new Scanner(System.in);

    /* -------------------------- Model classes -------------------------- */

    private static class Transaction {
        LocalDateTime timestamp;
        String type;
        double amount;
        double balanceAfter;

        Transaction(String type, double amount, double balanceAfter) {
            this.timestamp = LocalDateTime.now();
            this.type = type;
            this.amount = amount;
            this.balanceAfter = balanceAfter;
        }
    }

    private static class Account {
        private static final double DAILY_LIMIT = 5_000;
        private static final AtomicInteger NEXT_ACC_NO = new AtomicInteger(1001);

        private final int accNo;
        private final String holder;
        private double balance;
        private final List<Transaction> passbook = new ArrayList<>();

        /* Tracking daily withdrawal */
        private LocalDate lastWithdrawDate = null;
        private double withdrawnToday = 0;

        Account(String holder, double initial) {
            this.accNo = NEXT_ACC_NO.getAndIncrement();
            this.holder = holder;
            this.balance = initial;
            addTxn("OPEN", initial);
        }

        void deposit(double amt) {
            balance += amt;
            addTxn("DEPOSIT", amt);
        }

        boolean withdraw(double amt) {
            LocalDate today = LocalDate.now();
            if (!today.equals(lastWithdrawDate)) {
                withdrawnToday = 0;
                lastWithdrawDate = today;
            }

            if (amt > balance) {
                System.out.println("Insufficient balance.");
                return false;
            }
            if (withdrawnToday + amt > DAILY_LIMIT) {
                System.out.println("Daily withdrawal limit exceeded.");
                return false;
            }

            balance -= amt;
            withdrawnToday += amt;
            addTxn("WITHDRAW", amt);
            return true;
        }

        double getBalance() { return balance; }

        void printInfo() {
            System.out.printf("Account No : %d%nName       : %s%nBalance    : $%.2f%n", accNo, holder, balance);
        }

        void printPassbook() {
            printPassbook(LocalDateTime.MIN, LocalDateTime.MAX);
        }

        void printPassbook(LocalDateTime from, LocalDateTime to) {
            DateTimeFormatter fmt = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
            System.out.println("\n--- Passbook ---");
            passbook.stream()
                    .filter(t -> !t.timestamp.isBefore(from) && !t.timestamp.isAfter(to))
                    .forEach(t -> System.out.printf("%s | %8s | %10.2f | Bal: %.2f%n",
                            t.timestamp.format(fmt), t.type, t.amount, t.balanceAfter));
        }

        private void addTxn(String type, double amt) {
            passbook.add(new Transaction(type, amt, balance));
        }
    }

    /* --------------------------- Main driver --------------------------- */

    public static void main(String[] args) {
        Account account = null;
        while (true) {
            System.out.println("\n===== Banking System =====");
            System.out.println("1. Create account\n2. Deposit\n3. Withdraw\n4. Check balance\n5. Display account info\n6. Passbook print\n0. Exit");
            System.out.print("Select: ");
            int choice = readInt();

            if (choice == 0) break;

            if (account == null && choice != 1) {
                System.out.println("Please create an account first.");
                continue;
            }

            switch (choice) {
                case 1 -> account = createAccount();
                case 2 -> {
                    System.out.print("Deposit amount: $");
                    double amt = readDouble();
                    account.deposit(amt);
                }
                case 3 -> {
                    System.out.print("Withdraw amount: $");
                    double amt = readDouble();
                    account.withdraw(amt);
                }
                case 4 -> System.out.printf("Current balance: $%.2f%n", account.getBalance());
                case 5 -> account.printInfo();
                case 6 -> printPassbookMenu(account);
                default -> System.out.println("Invalid option.");
            }
        }
        System.out.println("Good-bye!");
    }

    private static Account createAccount() {
        System.out.print("Enter name: ");
        String name = scanner.nextLine();
        if (name.isEmpty()) name = scanner.nextLine(); // flush if earlier newline
        System.out.print("Initial deposit: $");
        double initial = readDouble();
        System.out.println("Account created successfully!");
        return new Account(name, initial);
    }

    private static void printPassbookMenu(Account account) {
        System.out.print("Print full passbook? (y/n): ");
        String yn = scanner.next();
        if (yn.equalsIgnoreCase("y")) {
            account.printPassbook();
        } else {
            System.out.print("From (yyyy-MM-dd HH:mm): ");
            LocalDateTime from = readDateTime();
            System.out.print("To   (yyyy-MM-dd HH:mm): ");
            LocalDateTime to = readDateTime();
            account.printPassbook(from, to);
        }
    }

    /* ------------------------- Helper methods ------------------------- */

    private static int readInt() {
        while (!scanner.hasNextInt()) {
            System.out.print("Invalid, re-enter: ");
            scanner.next();
        }
        return scanner.nextInt();
    }

    private static double readDouble() {
        while (!scanner.hasNextDouble()) {
            System.out.print("Invalid, re-enter: $");
            scanner.next();
        }
        return scanner.nextDouble();
    }

    private static LocalDateTime readDateTime() {
        DateTimeFormatter fmt = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
        while (true) {
            String s = scanner.next() + (scanner.hasNext() ? " " + scanner.next() : "");
            try {
                return LocalDateTime.parse(s, fmt);
            } catch (Exception e) {
                System.out.print("Invalid format, try again: ");
            }
        }
    }
}
